pkg/pullsync/pullsync.go

```diff
@@ -0,0 +1,332 @@
+			ctr++
+			bv.Set(i / swarm.HashSize)
+		}
+	}
+
+	wantMsg := &pb.Want{BitVector: bv.Bytes()}
+	if err = w.WriteMsgWithContext(ctx, wantMsg); err != nil {
+		return 0, fmt.Errorf("write want: %w", err)
+	}
+
+	// if ctr is zero, it means we don't want any chunk in the batch
+	// thus, the following loop will not get executed and the method
+	// returns immediately with the topmost value on the offer, which
+	// will seal the interval and request the next one
+
+	for ; ctr > 0; ctr-- {
+		var delivery pb.Delivery
+		if err = r.ReadMsgWithContext(ctx, &delivery); err != nil {
+			return 0, fmt.Errorf("read delivery: %w", err)
+		}
+
+		addr := swarm.NewAddress(delivery.Address)
+		if _, ok := wantChunks[addr.String()]; !ok {
+			return 0, ErrUnsolicitedChunk
+		}
+
+		delete(wantChunks, addr.String())
+		s.logger.Tracef("pull sync putting chunk %s", addr.String())
+		if err = s.storage.Put(ctx, storage.ModePutSync, swarm.NewChunk(addr, delivery.Data)); err != nil {
+			return 0, fmt.Errorf("delivery put: %w", err)
+		}
+	}
+	return offer.Topmost, nil
+}
+
+// handler handles an incoming request to sync an interval
+func (s *Syncer) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream) error {
+	w, r := protobuf.NewWriterAndReader(stream)
+	defer stream.Close()
+
+	var rn pb.GetRange
+	if err := r.ReadMsgWithContext(ctx, &rn); err != nil {
+		return fmt.Errorf("read get range: %w", err)
+	}
+	s.logger.Debugf("got range peer %s request %s", p.Address.String(), rn.String())
+
+	// make an offer to the upstream peer in return for the requested range
+	offer, addrs, err := s.makeOffer(ctx, rn)
+	if err != nil {
+		return fmt.Errorf("make offer: %w", err)
+	}
+	s.logger.Debugf("writing offer with context: %s", offer.String())
+
+	if err := w.WriteMsgWithContext(ctx, offer); err != nil {
+		return fmt.Errorf("write offer: %w", err)
+	}
+
+	// we don't have any hashes to offer in this range (the
+	// interval is empty). nothing more to do
+	if len(offer.Hashes) == 0 {
+		return nil
+	}
+
+	var want pb.Want
+	if err := r.ReadMsgWithContext(ctx, &want); err != nil {
+		return fmt.Errorf("read want: %w", err)
+	}
+
+	// empty bitvector implies downstream peer does not want
+	// any chunks (it has them already). mark chunks as synced
+	if len(want.BitVector) == 0 {
+		return s.setChunks(ctx, addrs...)
+	}
+
+	chs, err := s.processWant(ctx, offer, &want)
+	if err != nil {
+		return fmt.Errorf("process want: %w", err)
+	}
+
+	for _, v := range chs {
+		deliver := pb.Delivery{Address: v.Address().Bytes(), Data: v.Data()}
+		if err := w.WriteMsgWithContext(ctx, &deliver); err != nil {
+			return fmt.Errorf("write delivery: %w", err)
+		}
+	}
+
+	err = s.setChunks(ctx, addrs...)
+	if err != nil {
+		return err
+	}
+
+	time.Sleep(50 * time.Millisecond) //because of test, getting EOF w/o
+	return nil
+}
+
+func (s *Syncer) setChunks(ctx context.Context, addrs ...swarm.Address) error {
+	return s.storage.Set(ctx, storage.ModeSetSyncPull, addrs...)
```

Comment:
I dont thin we need ModeSetSyncPull any more. it was legacy

Reply:
i dont remember this was defined anywhere. can we iterate on this?